edgedb.go

```diff
@@ -54,29 +62,98 @@ func ConnectOne(ctx context.Context, opts Options) (*Conn, error) { // nolint:go
 
 // connectOne expectes a singleConn that has a nil net.Conn.
 func connectOne(ctx context.Context, opts *Options, conn *baseConn) error {
-	var d net.Dialer
-	netConn, err := d.DialContext(ctx, opts.network(), opts.address())
+	var (
+		d   net.Dialer
+		err error
+	)
+
+	conn.conn, err = d.DialContext(ctx, opts.network(), opts.address())
 	if err != nil {
-		return err
+		return fmt.Errorf("%v", err)
 	}
 
-	conn.conn = netConn
-	err = conn.connect(ctx, opts)
-	if err != nil {
-		return err
+	conn.writer = buff.NewWriter()
+	r := buff.NewReader()
+
+	if e := conn.setDeadline(ctx); e != nil {
+		return e
+	}
+
+	if e := conn.connect(r, opts); e != nil {
+		return e
 	}
 
+	conn.readerChan = make(chan *buff.Reader)
+	conn.releaseReader(r)
 	return nil
 }
 
+func (c *baseConn) setDeadline(ctx context.Context) error {
+	deadline, _ := ctx.Deadline()
+	if e := c.conn.SetDeadline(deadline); e != nil {
+		return fmt.Errorf("%v", e)
+	}
+
+	return nil
+}
+
+func (c *baseConn) acquireReader(ctx context.Context) (*buff.Reader, error) {
+	if e := c.conn.SetReadDeadline(time.Now()); e != nil {
```

Comment:
I initially implemented a background go routine that read from the socket in a tight loop. That started getting messy when it came to recovering from errors. This implementation starts a background reader when the `client.QueryX()` call ends and stops it when a new query begins. The down side to this is that the only reliable way to get the reader out of the background is to make it timeout. Do you think this is an acceptable compromise on complexity @1st1?

Reply:
> The down side to this is that the only reliable way to get the reader out of the background is to make it timeout.

Hm, this sounds really suboptimal. Is it possible to push some sentinel value to wake up the listener goroutine blocked on `buff.Reader.Next()`?

Reply:
Actually I'm not sure if it's possible to implement what I suggested.

> I initially implemented a background go routine that read from the socket in a tight loop. That started getting messy when it came to recovering from errors. 

Can you elaborate on the messy part? Because having a tight loop goroutine to read/parse messages and push signals to channels would make everything else easier.

Basically, you'd have a ReadBuffer where you'd feed data in that tight loop reader goroutine. Whenever there's a new complete message in that buffer you'd push `true` to the channel. 

This way, in say `Query()` you will just try reading directly from `ReadBuffer` to see if there's data in it. If not, you'd wait on the channel until the data is available.

And for asynchronous parsing you'd have yet another goroutine just listening on two channels: one for the ReadBuffer, and another would be used to stop it from the next `Query()` call.

The first goroutine (the one that populates ReadBuffer) can be stopped by closing the socket.

Reply:
I'll roll back to my original implementation and see if I can get that working.

This is not my original concern but loosely related: What happens when a query times out part way through reading data from the socket? The next read from the socket is going to return the remaining part of the timed out message(s). How does the client recover in this scenario? Close the connection and reconnect?

Reply:
> What happens when a query times out part way through reading data from the socket?

We will soon have an ability to enforce timeouts on the server side for every query. That way, if the query times out on the client side, the client ideally should wait a bit longer to try to read all messages until `SYNC`.  Or else it should just terminate the connection and open a new one.

Reply:
> I'll roll back to my original implementation and see if I can get that working.

Thanks! I think it should be doable; maybe some parts will be tricky but the overall solution in the end should be fast and intuitive.

Reply:
updated with background reader go routine